HierarchicalClustering/Communities.py

```diff
@@ -1,172 +1,236 @@
+    """
+    current_node = source_node
+    encountered_nodes = set()
+    path = ''
+    for step in range(max_path_length):
 
-    def _cluster_nodes_by_truncated_hitting_times(self, nodes_random_walk_data, threshold_hitting_time_difference):
-        """
-        Clusters a list of nodes into groups based on the truncated hitting
-        criterion as follows:
-
-        Let h_{j} be the average truncated hitting time of node v_{j}.
-        Nodes v_{j} are grouped into disjoint sets A_{k} such that:
-        for all v_{j} in A_{k} there exists a node v_{j'} in A_{k}
-        such that |h_{j} - h_{j'}| <= merge_threshold.
-        Ref: https://alchemy.cs.washington.edu/papers/kok10/kok10.pdf
-        """
+        next_node, next_edge = hypergraph.get_random_neighbor_and_edge_of_node(current_node)
+        path += str(next_edge.predicate) + ','
 
-        # sort the nodes in the hypergraph in increasing order of average hitting time
-        nodes = sorted(nodes_random_walk_data, key=lambda n: n.average_hitting_time)
-        current_hitting_time = nodes[0].average_hitting_time
-        distance_symmetric_clusters = []
-        distance_symmetric_single_nodes = []
-        distance_symmetric_cluster = []
-        for node in nodes:
-            if (node.average_hitting_time - current_hitting_time) < threshold_hitting_time_difference:
-                distance_symmetric_cluster.append(node)
-            else:
-                if len(distance_symmetric_cluster) == 1:
-                    distance_symmetric_single_nodes.append(distance_symmetric_cluster[0])
-                else:
-                    distance_symmetric_clusters.append(distance_symmetric_cluster.copy())
+        if next_node.name not in encountered_nodes:
+            nodes_random_walk_data[next_node.name].number_of_hits += 1
+            nodes_random_walk_data[next_node.name].add_path(path)
+            hitting_time = step + 1
+            nodes_random_walk_data[next_node.name].update_accumulated_hitting_time(hitting_time)
+            encountered_nodes.add(next_node.name)
 
-                distance_symmetric_cluster.clear()
-                distance_symmetric_cluster.append(node)
+        current_node = next_node
 
-            current_hitting_time = node.average_hitting_time
 
-        # append the last cluster to the list if it is not empty
-        if distance_symmetric_cluster is not None:
-            distance_symmetric_clusters.append(distance_symmetric_cluster)
+def get_close_nodes(nodes_random_walk_data, threshold_hitting_time: float):
+    return [node for node in nodes_random_walk_data.values() if node.average_hitting_time < threshold_hitting_time]
 
-        return distance_symmetric_single_nodes, distance_symmetric_clusters
 
-    def _cluster_nodes_by_js_divergence(self, node_random_walk_data, threshold_js_divergence, max_frequent_paths):
+def get_community(close_nodes: list[NodeRandomWalkData], node_types: set, config: dict):
+    single_nodes = []
```

Comment:
In Alchemy each source node is associated with a Community (a collection of single nodes and node clusters), so we are following their convention here. I think it would be confusing to call each cluster a community, since each of these 'communities' would then somehow need to be grouped and associated with a set single nodes.

Alternative names to Community?
Perhaps: 'NodeNeighborhood', 'NodeCommunity', ...

personally, I tend to think 'Community' is best.